src/app/WriteClient.cpp

```diff
@@ -247,29 +248,111 @@ CHIP_ERROR WriteClient::PutSinglePreencodedAttributeWritePayload(const chip::app
     return err;
 }
 
+CHIP_ERROR
+WriteClient::TryPutPreencodedAttributeWritePayloadIntoList(const chip::app::ConcreteDataAttributePath & attributePath,
+                                                           TLV::TLVReader & valueReader, bool & outChunkingNeeded,
+                                                           ListIndex & outEncodedItemCount)
+{
+
+    ReturnErrorOnFailure(EnsureListStarted(attributePath));
+
+    chip::TLV::TLVWriter * writer = nullptr;
+    VerifyOrReturnError((writer = GetAttributeDataIBTLVWriter()) != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    TLV::TLVWriter backupWriter;
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    outEncodedItemCount = 0;
+
+    while ((err = valueReader.Next()) == CHIP_NO_ERROR)
+    {
+        mWriteRequestBuilder.GetWriteRequests().Checkpoint(backupWriter);
```

Comment:
a comment explaining more here would help: we checkpoint `mWriteRequestBuilder`.GetWriteRequests` yet we write into `GetAttributeDataIBTLVWriter()` so they are interlinked and it is not immediately obvious why/how by just reading this code.

Reply:
@andy31415  What do you think of :
```suggestion
        // We essentially write List items to the AttributeDataIB, which is part of the AttributeDataIBs
        // Array in the WriteRequestMessage. As such, we checkpoint the outer AttributeDataIBs TLV writer
        // in case there's not enough space to encode a list item.
        mWriteRequestBuilder.GetWriteRequests().Checkpoint(backupWriter);
```

Reply:
I would explcitly call out that `GetAttributeDataIBTLVWriter` references this data. In my understanding (and looking for the comment to confirm or deny):

`mWriteRequestsBuilder` is sharing a single TLV writer among all objects. As such, `mWriteRequestsBuilder.GetWriteRequests()` as well as `*writer` are essentionally the same thing.

However if I read that correctly, why are we not using 

```cpp
writer->Checkpoint(backupWriter);
```

here? That makes it super obvious that we "try to write the element and if we fail, we roll back". Why are we going through different objects? The comment should explain  why. Otherwise change the code to `writer->Checkpoint` instead of `mWriteRequestBuilder.GetWriteRequests().Checkpoint`

Reply:
you are totally right, we are indeed sharing a single TLV Writer. As discussed, I Made the code clearer by choosing more appropriate names
```
        // We essentially write List items to the Data Container within the first AttributeDataIB. We checkpoint the AttributeDataIB
        // Builder in case there's not enough space to encode a list item.
        attributeDataIB.Checkpoint(backupWriter);
        err = attributeDataIBWriter->CopyElement(TLV::AnonymousTag(), valueReader);

        if (err == CHIP_ERROR_NO_MEMORY || err == CHIP_ERROR_BUFFER_TOO_SMALL)
        {
            attributeDataIB.Rollback(backupWriter);
```